Multi-dimensional flow model implementation based on SAM implementation

lingzou opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->
This pull request includes multi-dimensional flow model implementation, both Navier-Stokes model and porous-medium model, based on existing SAM implementation

## Reason
<!--Why do you need this feature or what is the enhancement?-->
The SAM multi-dimensional flow model is a generic CFD-like implementation which should be made available to public.

## Design
<!--A concise description (design) of the enhancement.-->
Based on existing SAM implementation.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
Users have more choices when selecting multi-dimensional flow model implementation.

Closes #23121

lingzou-anl committed:
Add multi-dimensional flow model kernerls based on SAM implementations
Refs #22917

lingzou-anl committed:
Add multi-dimensional flow model bcs based on SAM implementations
Refs #22917

lingzou-anl committed:
Fix compiling issues:
Use SinglePhaseFluidProperties, instead of PTEquationOfState
Remove inheritence from FluidPressureTimeDerivative

Refs #22917

lingzou-anl committed:
Add MDFluidMaterial

Refs #22917

lingzou-anl committed:
Add lid driven problem (same order variables + PSPG/SUPG) into regression tests
Refs #22917

lingzou-anl committed:
Make the existing kernels more completed by consolidating other kernels
Refs #22917

lingzou-anl committed:
Add porous medium flow model and tests
Refs #22917

lingzou commented:
@lindsayad There are a couple of minor things to be fixed but the pull request is ready for review.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1259176/) on 857bb38 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22671/site)

This comment will be updated on new commits.

lingzou-anl committed:
minor fix in MDFluidEnergyBC.C; Refs #22917

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1264124/) on 6246c8c wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>d653ad</th>
      <th colspan=4>#22671 6246c8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.56%</td>
      <td>84.49%</td>
      <td>-1.07%</td>
      <td>70.28%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>9781</td>
      <td>10384</td>
      <td>+603</td>
      <td>603</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1651</td>
      <td>1906</td>
      <td>+255</td>
      <td>255</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22671/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22671/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22671/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 70.28% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented:
@lingzou is it ok for me to push some commits to your branch? I see some style things I'd like to try to change like `map_var_number` to `mapVarNumber` just to be more consistent with naming in the repo

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidEnergyBC.h:
```diff
@@ -0,0 +1,43 @@
+#include "MDFluidIntegratedBCBase.h"
+#include "Function.h"
+
+/* An integral BC for the energy (temperature) equation  */
```
```suggestion
/**
  * An integral BC for the energy (temperature) equation 
  */
```

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidEnergyBC.h:
```diff
@@ -0,0 +1,43 @@
+  virtual ~MDFluidEnergyBC() {}
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
```
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:
```diff
@@ -0,0 +1,42 @@
+#include "NodalBC.h"
+#include "Function.h"
+
+/* A conditional Dirichlet BC for the energy (temperature) equation  */
```
```suggestion
/**
  * A conditional Dirichlet BC for the energy (temperature) equation  
  */
```

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:
```diff
@@ -0,0 +1,42 @@
+  virtual ~MDFluidEnergyDirichletBC() {}
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
```
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:
```diff
@@ -0,0 +1,42 @@
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
+
+  bool shouldApply();
```
```suggestion
  bool shouldApply() override;
```

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidIntegratedBCBase.h:
```diff
@@ -0,0 +1,58 @@
+
+  // Helper function for mapping Moose variable numberings into
+  // the "canonical" numbering for the porous medium equations.
+  unsigned map_var_number(unsigned var);
```
```suggestion
  unsigned mapVarNumber(unsigned var);
```

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidMassBC.h:
```diff
@@ -0,0 +1,33 @@
+#include "MDFluidIntegratedBCBase.h"
+#include "Function.h"
+
+/* A specific BC for the mass (pressure) equation  */
```
```suggestion
/**
  * A specific BC for the mass (pressure) equation 
  */
```

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidMassBC.h:
```diff
@@ -0,0 +1,33 @@
+  virtual ~MDFluidMassBC() {}
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
```
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

lindsayad commented on modules/navier_stokes/include/bcs/MDFluidMomentumBC.h:
```diff
@@ -0,0 +1,43 @@
+  MDFluidMomentumBC(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
```
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

lindsayad commented on modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h:
```diff
@@ -0,0 +1,43 @@
+
+#include "Function.h"
+
+// Forward Declarations
+class MDFluidEnergyKernel;
+
```
```suggestion
```

lindsayad commented on modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h:
```diff
@@ -0,0 +1,43 @@
+// Forward Declarations
+class MDFluidEnergyKernel;
+
+// The spatial part of the 3D energy conservation for fluid flow
```
```suggestion
/**
  * The spatial part of the 3D energy conservation for fluid flow
  */
```

lindsayad commented on modules/navier_stokes/include/kernels/MDFluidKernelBase.h:
```diff
@@ -0,0 +1,85 @@
+
+protected:
+  Real velocityDiv() { return _grad_u_vel[_qp](0) + _grad_v_vel[_qp](1) + _grad_w_vel[_qp](2); }
+  RealVectorValue velocityDot()
+  {
+    RealVectorValue vec_vel_dot(_u_vel_dot[_qp], _v_vel_dot[_qp], _w_vel_dot[_qp]);
+    return vec_vel_dot;
+  }
```
In general I try to encourage people (including myself who's often a guilty party) to move multi-line definitions out of the class definition, e.g. we could make this inline in the header outside of the class definition

lindsayad commented on modules/navier_stokes/include/kernels/MDFluidKernelBase.h:
```diff
@@ -0,0 +1,85 @@
+  const MaterialProperty<RealTensorValue> & _inertia_resistance_coeff;
+  const MaterialProperty<RealTensorValue> & _viscous_resistance_coeff;
+
+  // Helper function for mapping Moose variable numberings into
+  // the "canonical" numbering for the porous medium equations.
+  unsigned map_var_number(unsigned var);
```
```suggestion
  /**
    * Helper function for mapping Moose variable numberings into
    * the "canonical" numbering for the porous medium equations.
    */
  unsigned int mapVarNumber(unsigned int var) const;
```

lindsayad commented on modules/navier_stokes/include/kernels/MDFluidKernelStabilization.h:
```diff
@@ -0,0 +1,33 @@
+
+#include "MDFluidKernelBase.h"
+
+/* Base class for stabilization kernels. */
```
```suggestion
/**
  * Base class for stabilization kernels.
  */
```

lindsayad commented on modules/navier_stokes/include/kernels/MDFluidMassKernel.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "MDFluidKernelStabilization.h"
+
+// The spatial part of the 3D mass conservation for fluid flow
```
```suggestion
/**
  * The spatial part of the 3D mass conservation for fluid flow
  */
```

lindsayad commented on modules/navier_stokes/include/kernels/MDFluidMomentumKernel.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "MDFluidKernelStabilization.h"
+
+// The spatial part of the 3D momentum conservation for fluid flow
```
```suggestion
/**
  * The spatial part of the 3D momentum conservation for fluid flow
  */
```

lindsayad commented on modules/navier_stokes/include/kernels/PMFluidPressureTimeDerivative.h:
```diff
@@ -0,0 +1,36 @@
+#include "TimeKernel.h"
+#include "SinglePhaseFluidProperties.h"
+
+// The transient term of the porous-media mass conservation equation
```
```suggestion
/**
  * The transient term of the porous-media mass conservation equation
  */
```

lindsayad commented on modules/navier_stokes/include/materials/MDFluidMaterial.h:
```diff
@@ -0,0 +1,76 @@
+#include "Material.h"
+#include "SinglePhaseFluidProperties.h"
+
+/* Fluid materials for 3D fluid model */
```
```suggestion
/**
  * Fluid materials for 3D fluid model
  */
```

lindsayad commented on modules/navier_stokes/src/auxkernels/NSDensityAux.C:
```diff
@@ -0,0 +1,39 @@
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NSDensityAux.h"
```
```suggestion
#include "NSDensityAux.h"
#include "NS.h"
```

lindsayad commented on modules/navier_stokes/src/auxkernels/NSDensityAux.C:
```diff
@@ -0,0 +1,39 @@
+  InputParameters params = AuxKernel::validParams();
+
+  // Coupled variables
+  params.addRequiredCoupledVar("pressure", "Coupled fluid pressure variable");
+  params.addRequiredCoupledVar("temperature", "Coupled fluid temperature variable");
```
```suggestion
  params.addRequiredCoupledVar(NS::pressure, "Coupled fluid pressure variable");
  params.addRequiredCoupledVar(NS::temperature, "Coupled fluid temperature variable");
```

lindsayad reviewed (dismissed):
Some preliminary review through `include`. @lingzou I'd be happy to push commits to address these kind of cosmetic changes to help reduce time burden on you. Just let me know

lingzou commented:
> 

@lindsayad Please go ahead do the necessary changes. Let me know if I have to do anything on my end, e.g., giving you access to the repo, things like that.

lindsayad commented:
Sounds good!

lindsayad committed:
Apply suggestions from code review

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1289958/) on 73cf6a8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22671/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22671/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fe98211c5cb7fddf763b8db977bc5a558912402e`

lindsayad committed:
git sed map_var_number mapVarNumber

lindsayad committed:
Documentation stub pages

lindsayad committed:
git mv porous_medium pins/channel-flow

lindsayad committed:
Add class descriptions

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1291345/) on 866462f : invalidated by @lindsayad

lindsayad committed:
Make stubs warnings

lindsayad commented:
@lingzou would it be possible to add a couple more tests? Currently it looks like the following objects aren't used:
- FluidWallMomentumBC
- MDFluidEnergyBC
- MDFluidEnergyDirichletBC
- MDFluidMomentumBC
- MDMomentumFreeSlipBC
I looked at the possibility of porting some of the existing tests in SAM that use these objects but it looks like they use the non-porous time derivative kernels so I would have to port those as well ... and it looks like some of those time derivative kernels use EOS APIs that don't have drop-in replacements with `SinglePhaseFluidProperties`, e.g. `enthalpy_from_T`. We have `h_from_p_T` but I didn't want to go too far down the rabbit hole of porting those kernels and changing to `SinglePhaseFluidProperties` and changing to `p_T` lookups ... as you may very well want to continue to use objects like `PBSodiumEquationOfState`

lingzou commented:
> @lingzou would it be possible to add a couple more tests? Currently it looks like the following objects aren't used:
> 
> * FluidWallMomentumBC
> * MDFluidEnergyBC
> * MDFluidEnergyDirichletBC
> * MDFluidMomentumBC
> * MDMomentumFreeSlipBC
> 
> I looked at the possibility of porting some of the existing tests in SAM that use these objects but it looks like they use the non-porous time derivative kernels so I would have to port those as well ... and it looks like some of those time derivative kernels use EOS APIs that don't have drop-in replacements with `SinglePhaseFluidProperties`, e.g. `enthalpy_from_T`. We have `h_from_p_T` but I didn't want to go too far down the rabbit hole of porting those kernels and changing to `SinglePhaseFluidProperties` and changing to `p_T` lookups ... as you may very well want to continue to use objects like `PBSodiumEquationOfState`

@lindsayad thanks for keeping working on the PR. It is probably the best for me to add/port additional tests to cover these BCs. It shouldn't take long to do so, and we will let Elia understand that we are still working on it.

lindsayad commented:
That would be great, thanks!

lingzou committed:
Add additional regression tests

lindsayad committed:
Allow warnings for FDP in parallel

lindsayad committed:
Update root module documentation

- Add `MD` and `PM` to the prefix set
- Add note stating that `INS` and `MD` will eventually be consolidated

moosebuild commented:
Job [No optional](https://civet.inl.gov/job/1308941/) on 047e01b : invalidated by @lindsayad

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1308907/) on 047e01b : invalidated by @lindsayad

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/1308931/) on 047e01b : invalidated by @lindsayad

lindsayad commented:
Ok @GiudGiud, @grmnptr I have commits on this PR so for SQA I cannot approve/merge it. Can one of you guys review this? Note that we cannot change the functionality of these objects as we need to maintain SAM compatibility and I'm not going to spend time on consolidation of the FE objects before this is merged per Elia's request

GiudGiud reviewed (commented):
NSDensityAux should be deleted we have FluidDensityAux in the FP module
FluidWallMomentumBC is missing a prefix

Please link issues for:
- missing docs & docstrings
- issues attached to the original code
- missing Jacobian terms for that test that uses the FDP

Can we rename MD and PM? To INSFE and PINSFE most likely?

lindsayad commented:
> Can we rename MD and PM? To INSFE and PINSFE most likely?

We could potentially do aliases. If we renamed we would break SAM

GiudGiud commented:
Aliasing seems fine for a first step.

lindsayad commented:
Ok aliasing doesn't seem like the correct solution. That only makes sense if you want the name in the `Registry` (and the name usable in the input file) to be different than the actual class name. It's not to support multiple allowable names in an input file. So we could deprecate the current names in favor of new names ... then both names would be usable in the input file but this would be denoting that we essentially want our users to adopt a new name eventually. @lingzou is this something that you think is acceptable?

lingzou commented:
> ... that we essentially want our users to adopt a new name eventually. @lingzou is this something that you think is acceptable?

@lindsayad Yes, that will be fine, as long as we provide a time window for end users to adopt the new syntax/names. [I am at-ing @lingzou-anl so my ANL email receives update].

lindsayad committed:
git mv MD -> INSFE, PM -> PINSFE

lindsayad committed:
git sed MD -> INSFE, PM -> PINSFE

lindsayad committed:
Remove NSDensityAux

lindsayad committed:
Add rename macros for SAM backwards compatibility

lindsayad committed:
Update index doc

GiudGiud commented on modules/navier_stokes/doc/content/source/materials/MDFluidMaterial.md:
```diff
@@ -0,0 +1,23 @@
+# MDFluidMaterial
```
rename?

GiudGiud commented on modules/navier_stokes/include/bcs/INSFEFluidWallMomentumBC.h:
```diff
@@ -0,0 +1,29 @@
+
+#include "INSFEFluidIntegratedBCBase.h"
+
+class INSFEFluidWallMomentumBC : public INSFEFluidIntegratedBCBase
```
docstring

GiudGiud commented on modules/navier_stokes/include/bcs/INSFEFluidMomentumBC.h:
```diff
@@ -0,0 +1,43 @@
+#include "INSFEFluidIntegratedBCBase.h"
+#include "Function.h"
+
+class INSFEFluidMomentumBC : public INSFEFluidIntegratedBCBase
```
same

GiudGiud commented on modules/navier_stokes/include/kernels/PINSFEFluidTemperatureTimeDerivative.h:
```diff
@@ -0,0 +1,32 @@
+#include "TimeDerivative.h"
+#include "SinglePhaseFluidProperties.h"
+
+class PINSFEFluidTemperatureTimeDerivative : public TimeDerivative
```
docstring

GiudGiud commented on modules/navier_stokes/include/kernels/PINSFEFluidVelocityTimeDerivative.h:
```diff
@@ -0,0 +1,32 @@
+#include "TimeDerivative.h"
+#include "SinglePhaseFluidProperties.h"
+
+class PINSFEFluidVelocityTimeDerivative : public TimeDerivative
```
same

GiudGiud commented on modules/navier_stokes/include/materials/GenericPorousMediumMaterial.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "MDFluidMaterial.h"
+
+class GenericPorousMediumMaterial : public MDFluidMaterial
```
GenericPINSFEMaterial ?

GiudGiud commented on modules/navier_stokes/doc/content/modules/navier_stokes/index.md:
```diff
@@ -27,6 +27,8 @@ Here we give a brief tabular summary of the Navier-Stokes implementations:
 | ------     | --------   | -------------------- | ------------------ | ----------------  | ------ | ------------------                |
 | INS        | Hand-coded | incompressible       | None               | Not porous        | CGFE   | SUPG                              |
 | INSAD      | AD         | incompressible       | Smagorinsky        | Not porous        | CGFE   | SUPG                              |
+| INSFE      | Hand-coded | incompressible       | None               | Not porous        | CGFE   | SUPG                              |
+| PINSFE     | Hand-coded | incompressible       | None               | porous            | CGFE   | SUPG                              |
```
both have mixing length & seemingly others, see the MDFluidMaterial

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/pins/expansion-channel/tests:
```diff
@@ -0,0 +1,12 @@
+    exodiff = 'expansion-channel-slip-wall_out.e'
+    abs_zero = 1e-6
+    issues = '#23121'
+    requirement = 'We shall be able to solve a porous flow using slip-wall boundary condition.'
```
```suggestion
    requirement = 'The system shall be able to solve a porous flow using slip-wall boundary condition.'
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:
```diff
@@ -0,0 +1,27 @@
+    exodiff = 'pm_friction_out.e'
+    abs_zero = 1e-6
+    issues = '#23121'
+    requirement = 'We shall be able to solve a porous flow with pressure drop due to viscous and inertia frictions.'
```
```suggestion
    requirement = 'The system shall be able to solve a porous flow with pressure drop due to viscous and inertia frictions.'
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:
```diff
@@ -0,0 +1,27 @@
+    exodiff = 'pm_heat_source_out.e'
+    abs_zero = 1e-6
+    issues = '#23121'
+    requirement = 'We shall be able to solve a porous flow with internal heating source.'
```
```suggestion
    requirement = 'The system shall be able to solve a porous flow with internal heating source.'
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:
```diff
@@ -0,0 +1,27 @@
+    exodiff = 'pm_reverse_flow_out.e'
+    abs_zero = 1e-6
+    issues = '#23121'
+    requirement = 'We shall be able to solve a porous flow when reverse flow happens.'
```
```suggestion
    requirement = 'The system shall be able to solve a porous flow when reverse flow happens.'
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/lid_driven/tests:
```diff
@@ -8,6 +8,14 @@
+    exodiff = 'lid_driven_md_out.e'
+    custom_cmp = 'lid_driven.cmp'
+    issues = '#23121'
+    requirement = 'We shall be able to solve a canonical lid-driven problem using same order variables and the PSPG/SUPG stabilization scheme.'
```
```suggestion
    requirement = 'The system shall be able to solve a canonical lid-driven problem using same order variables and the PSPG/SUPG stabilization scheme.'
```

GiudGiud reviewed (commented):
ok let s get this in after these. 

In terms of coverage we could add two turbulence tests. 
I m ok with adding them later
Same for all object docs

lindsayad committed:
Address Guillaume review

- 'We' -> 'The system'
- Some missing doc strings
- Rename materials to INSFE and PINSFE
- Some cleanup of method names in the materials to match MOOSE syntax

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Correct the drho_dt calculations; use pressure_dot

moosebuild commented:
Job [Conda (Ubuntu)](https://civet.inl.gov/job/1324227/) on 0f8f5e9 : invalidated by @GiudGiud

fluke

GiudGiud reviewed (dismissed):
5 last comments can be done in a follow up they are minor

lindsayad commented:
Sorry @GiudGiud I pushed another commit with my airplane work, so will need another review

lindsayad commented:
I tried to address the other 5 conversations
